File: backend/src/domain/financing/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::financing::schema::BusinessPlan,
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct FinancingDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl FinancingDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_business_plan(&self, id: i64) -> Result<BusinessPlan, String> {
        self.db_dao
            .get_business_plan(id)
            .await
            .map(|record| BusinessPlan {
                id: record.id,
                title: record.title,
                content: record.content,
                industry: record.industry,
                region: record.region,
                financing_amount: record.financing_amount,
                company_size: record.company_size,
                created_at: record.created_at,
            })
            .map_err(|e| e.to_string())
    }

    pub async fn get_business_plans(&self, ids: Vec<i64>) -> Result<Vec<BusinessPlan>, String> {
        self.db_dao
            .get_business_plans(ids)
            .await
            .map(|records| {
                records
                    .into_iter()
                    .map(|record| BusinessPlan {
                        id: record.id,
                        title: record.title,
                        content: record.content,
                        industry: record.industry,
                        region: record.region,
                        financing_amount: record.financing_amount,
                        company_size: record.company_size,
                        created_at: record.created_at,
                    })
                    .collect()
            })
            .map_err(|e| e.to_string())
    }

    pub async fn create_business_plan(
        &self,
        title: &str,
        content: &str,
        industry: &str,
        region: &str,
        financing_amount: f64,
        company_size: &str,
    ) -> Result<i64, String> {
        self.db_dao
            .create_business_plan(
                title,
                content,
                industry,
                region,
                financing_amount,
                company_size,
            )
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_business_plan(
        &self,
        id: i64,
        title: Option<String>,
        content: Option<String>,
        industry: Option<String>,
        region: Option<String>,
        financing_amount: Option<f64>,
        company_size: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_business_plan(
                id,
                title,
                content,
                industry,
                region,
                financing_amount,
                company_size,
            )
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn delete_business_plan(&self, id: i64) -> Result<i64, String> {
        self.db_dao
            .delete_business_plan(id)
            .await
            .map_err(|e| e.to_string())
    }
}

File: backend/domain/financing/domain.go
package financing

import (
	"backend/datasource/dbdao"
	"context"
	"errors"
)

type FinancingDomain struct {
	DB *dbdao.DB
}

func NewFinancingDomain(db *dbdao.DB) *FinancingDomain {
	return &FinancingDomain{DB: db}
}

// GetBusinessPlan 获取商业计划
func (d *FinancingDomain) GetBusinessPlan(ctx context.Context, id int64) (*BusinessPlan, error) {
	plan, err := d.DB.GetBusinessPlan(ctx, id)
	if err != nil {
		return nil, err
	}

	return &BusinessPlan{
		ID:              plan.ID,
		Title:           plan.Title,
		Content:         plan.Content,
		Industry:        plan.Industry,
		Region:          plan.Region,
		FinancingAmount: plan.FinancingAmount,
		CompanySize:     plan.CompanySize,
		CreatedAt:       &plan.CreatedAt,
		UpdatedAt:       &plan.UpdatedAt,
	}, nil
}

// GetBusinessPlans 批量获取商业计划
func (d *FinancingDomain) GetBusinessPlans(ctx context.Context, ids []int64) ([]*BusinessPlan, error) {
	if len(ids) == 0 {
		return []*BusinessPlan{}, nil
	}

	plans, err := d.DB.GetBusinessPlans(ctx, ids)
	if err != nil {
		return nil, err
	}

	result := make([]*BusinessPlan, len(plans))
	for i, plan := range plans {
		result[i] = &BusinessPlan{
			ID:              plan.ID,
			Title:           plan.Title,
			Content:         plan.Content,
			Industry:        plan.Industry,
			Region:          plan.Region,
			FinancingAmount: plan.FinancingAmount,
			CompanySize:     plan.CompanySize,
			CreatedAt:       &plan.CreatedAt,
			UpdatedAt:       &plan.UpdatedAt,
		}
	}

	return result, nil
}

// CreateBusinessPlan 创建商业计划
func (d *FinancingDomain) CreateBusinessPlan(
	ctx context.Context,
	title, content, industry, region, companySize string,
	financingAmount float64,
) (*BusinessPlan, error) {
	if title == "" || content == "" || industry == "" || region == "" {
		return nil, errors.New("missing required fields")
	}

	plan := &dbdao.BusinessPlan{
		Title:           title,
		Content:         content,
		Industry:        industry,
		Region:          region,
		FinancingAmount: financingAmount,
		CompanySize:     companySize,
	}

	created, err := d.DB.CreateBusinessPlan(ctx, plan)
	if err != nil {
		return nil, err
	}

	return &BusinessPlan{
		ID:              created.ID,
		Title:           created.Title,
		Content:         created.Content,
		Industry:        created.Industry,
		Region:          created.Region,
		FinancingAmount: created.FinancingAmount,
		CompanySize:     created.CompanySize,
		CreatedAt:       &created.CreatedAt,
		UpdatedAt:       &created.UpdatedAt,
	}, nil
}

// UpdateBusinessPlan 更新商业计划
func (d *FinancingDomain) UpdateBusinessPlan(
	ctx context.Context,
	id int64,
	title, content, industry, region, companySize *string,
	financingAmount *float64,
) (*BusinessPlan, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}

	// 构建更新对象，只包含非nil字段
	plan := &dbdao.BusinessPlan{
		ID: id,
	}

	if title != nil {
		plan.Title = *title
	}
	if content != nil {
		plan.Content = *content
	}
	if industry != nil {
		plan.Industry = *industry
	}
	if region != nil {
		plan.Region = *region
	}
	if companySize != nil {
		plan.CompanySize = *companySize
	}
	if financingAmount != nil {
		plan.FinancingAmount = *financingAmount
	}

	updated, err := d.DB.UpdateBusinessPlan(ctx, plan)
	if err != nil {
		return nil, err
	}

	return &BusinessPlan{
		ID:              updated.ID,
		Title:           updated.Title,
		Content:         updated.Content,
		Industry:        updated.Industry,
		Region:          updated.Region,
		FinancingAmount: updated.FinancingAmount,
		CompanySize:     updated.CompanySize,
		CreatedAt:       &updated.CreatedAt,
		UpdatedAt:       &updated.UpdatedAt,
	}, nil
}

// DeleteBusinessPlan 删除商业计划
func (d *FinancingDomain) DeleteBusinessPlan(ctx context.Context, id int64) error {
	if id <= 0 {
		return errors.New("invalid id")
	}
	return d.DB.DeleteBusinessPlan(ctx, id)
}
